File: javascript/test/parser.test.js
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2025 Ian Thomas

import {loadTestFile} from '../test/testUtils.js';
import {strict as assert} from 'assert';
import {ExpressionParser} from "../src/parser.js";

describe('ExpressionParser', () => {

  describe('Simple', () => {
    it('should match', () => {

      let parser = new ExpressionParser();
      let expression = parser.parse("get_name()=='fred' and counter>0 and 5/5.0!=0");

      let context = {
        "get_name": () => "fred",
        "counter": 1
      };

      let result = expression.evaluate(context);

      assert.equal(result, true);
    });
  });

  describe('Specificity', () => {
    it('should match', () => {

      let parser = new ExpressionParser();
      let expression = parser.parse("get_name()=='fred' and counter>0 and 5/5.0!=0");
      assert.equal(expression.specificity, 2);
      expression = parser.parse("get_name()=='fred' and counter>0 and (5/5.0!=0 or true)");
      assert.equal(expression.specificity, 3);
      expression = parser.parse("true");
      assert.equal(expression.specificity, 0);
    });
  });

  describe('MatchingOutput', () => {
    it('should match', () => {

      const source = loadTestFile('Parse.txt');

      const lines = source.split(/\r?\n/);

      const context = {
        C: 15,
        D: false,
        get_name: () => "fred",
        end_func: () => true,
        whisky: (id, n) => String(n) + "whisky_" + id,
        counter: 1,
      };
    
      const parser = new ExpressionParser();
    
      let processed_lines = [];
      for (const line of lines) {

        if (line.startsWith("//")) {
          processed_lines.push(line);
          continue;
        }
    
        processed_lines.push(`"${line}"`);
        try {

          const node = parser.parse(line);
          processed_lines.push(node.dump_structure());

          let dump_eval = [];
          node.evaluate(context, dump_eval);
          processed_lines.push(dump_eval.join("\n"));

        } catch (e) {
          processed_lines.push(`${e.message}`);
        }
        processed_lines.push("");
      }
    
      const output = processed_lines.join("\n");

      const match = loadTestFile('Parse-Output.txt');
      assert.equal(match, output);
    });
  });

});

File: csharp/ExpressionParser.Tests/ParserTest.cs
// This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
// Copyright (c) 2025 Ian Thomas

namespace FountainTools.Tests;
using System.IO;
using ExpressionParser;

public class ParserTest
{
    private string loadTestFile(string fileName) {
        return File.ReadAllText("../../../../../tests/"+fileName);
    }

    [Fact]
    public void Simple()
    {
        var parser = new Parser();
        var expression = parser.Parse("get_name()=='fred' and counter>0 and 5/5.0!=0");

        var context = new Dictionary<string, object>
        {
            { "get_name", new Func<string>(() => "fred") },
            { "counter", 1 }
        };

        var result = expression.Evaluate(context);

        Assert.Equal(true, result);
    }

    /*[Fact]
    public void Scratch()
    {
        string source = loadTestFile("Scratch.fountain");
        string match = loadTestFile("Scratch.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void SceneHeading()
    {
        string source = loadTestFile("SceneHeading.fountain");
        string match = loadTestFile("SceneHeading.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionMerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Merged.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void ActionUnmerged()
    {
        string source = loadTestFile("Action.fountain");
        string match = loadTestFile("Action-Unmerged.txt");

        Parser fp = new Parser();
        fp.MergeActions = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Character()
    {
        string source = loadTestFile("Character.fountain");
        string match = loadTestFile("Character.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueMerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Merged.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void DialogueUnmerged()
    {
        string source = loadTestFile("Dialogue.fountain");
        string match = loadTestFile("Dialogue-Unmerged.txt");

        Parser fp = new Parser();
        fp.MergeDialogue = false;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Parenthetical()
    {
        string source = loadTestFile("Parenthetical.fountain");
        string match = loadTestFile("Parenthetical.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Lyrics()
    {
        string source = loadTestFile("Lyrics.fountain");
        string match = loadTestFile("Lyrics.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Transition()
    {
        string source = loadTestFile("Transition.fountain");
        string match = loadTestFile("Transition.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void TitlePage()
    {
        string source = loadTestFile("TitlePage.fountain");
        string match = loadTestFile("TitlePage.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void PageBreak()
    {
        string source = loadTestFile("PageBreak.fountain");
        string match = loadTestFile("PageBreak.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void LineBreaks()
    {
        string source = loadTestFile("LineBreaks.fountain");
        string match = loadTestFile("LineBreaks.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Notes()
    {
        string source = loadTestFile("Notes.fountain");
        string match = loadTestFile("Notes.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

    [Fact]
    public void Boneyards()
    {
        string source = loadTestFile("Boneyards.fountain");
        string match = loadTestFile("Boneyards.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void Sections()
    {
        string source = loadTestFile("Sections.fountain");
        string match = loadTestFile("Sections.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void UTF8()
    {
        string source = loadTestFile("UTF8.fountain");
        string match = loadTestFile("UTF8.txt");

        Parser fp = new Parser();
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }

   [Fact]
    public void Tags()
    {
        string source = loadTestFile("Tags.fountain");
        string match = loadTestFile("Tags.txt");

        Parser fp = new Parser();
        fp.UseTags = true;
        fp.AddText(source);

        string output = fp.Script.Dump();
        //Console.WriteLine(output);
        Assert.Equal(match, output);
    }*/
}
